xmppserver/src/main/java/org/jivesoftware/openfire/session/LocalSession.java

```diff
@@ -135,16 +142,26 @@ public LocalSession(String serverName, Connection connection, StreamID streamID,
      * @return true if session detached
      */
     public boolean isDetached() {
-        return this.conn == null;
+        lock.readLock().lock();
+        try {
+            return this.conn == null;
+        }finally {
+            lock.readLock().unlock();
+        }
     }
 
     /**
      * Set the session to detached mode, indicating that the underlying connection
      * has been closed.
      */
     public void setDetached() {
-        this.sessionManager.addDetached(this);
-        this.conn = null;
+        lock.writeLock().lock();
```

Comment:
Not sure this lock achieves much either. It does mean the object can be in state of detached with a conn, but don't see how that can have any side effects.

Reply:
I think it can have side effects with the reattach method. So line 179 (https://github.com/igniterealtime/Openfire/pull/1427/files#diff-ea9a45ac8aa08fccbc303ffc0f705569R179) should also be in the lock.

The lock should ensure, that there is no other operation to which do a operation on the connection while it is set to null or another reference.

Reply:
what about this?